Index: source/org/jivesoftware/smack/provider/ProviderManager.java
===================================================================
--- source/org/jivesoftware/smack/provider/ProviderManager.java	(revision 12558)
+++ source/org/jivesoftware/smack/provider/ProviderManager.java	(working copy)
@@ -132,6 +132,7 @@
     public static synchronized ProviderManager getInstance() {
         if (instance == null) {
             instance = new ProviderManager();
+            instance.initialize();
         }
         return instance;
     }
@@ -430,9 +431,4 @@
         }
         return loaders.toArray(new ClassLoader[loaders.size()]);
     }
-
-    private ProviderManager() {
-        super();
-        initialize();
-    }
-}
\ No newline at end of file
+}
