Index: source/org/jivesoftware/smackx/packet/DiscoverInfo.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- source/org/jivesoftware/smackx/packet/DiscoverInfo.java	(revision 12975)
+++ source/org/jivesoftware/smackx/packet/DiscoverInfo.java	(revision )
@@ -21,6 +21,7 @@
 package org.jivesoftware.smackx.packet;
 
 import org.jivesoftware.smack.packet.IQ;
+import org.jivesoftware.smack.util.StringUtils;
 
 import java.util.Collections;
 import java.util.Iterator;
@@ -135,7 +136,7 @@
         buf.append("<query xmlns=\"http://jabber.org/protocol/disco#info\"");
         if (getNode() != null) {
             buf.append(" node=\"");
-            buf.append(getNode());
+            buf.append(StringUtils.escapeForXML(getNode()));
             buf.append("\"");
         }
         buf.append(">");
@@ -222,10 +223,10 @@
 
         public String toXML() {
             StringBuilder buf = new StringBuilder();
-            buf.append("<identity category=\"").append(category).append("\"");
-            buf.append(" name=\"").append(name).append("\"");
+            buf.append("<identity category=\"").append(StringUtils.escapeForXML(category)).append("\"");
+            buf.append(" name=\"").append(StringUtils.escapeForXML(name)).append("\"");
             if (type != null) {
-                buf.append(" type=\"").append(type).append("\"");
+                buf.append(" type=\"").append(StringUtils.escapeForXML(type)).append("\"");
             }
             buf.append("/>");
             return buf.toString();
@@ -262,8 +263,8 @@
 
         public String toXML() {
             StringBuilder buf = new StringBuilder();
-            buf.append("<feature var=\"").append(variable).append("\"/>");
+            buf.append("<feature var=\"").append(StringUtils.escapeForXML(variable)).append("\"/>");
             return buf.toString();
         }
     }
-}
\ No newline at end of file
+}
Index: source/org/jivesoftware/smackx/packet/DiscoverItems.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- source/org/jivesoftware/smackx/packet/DiscoverItems.java	(revision 12975)
+++ source/org/jivesoftware/smackx/packet/DiscoverItems.java	(revision )
@@ -21,6 +21,7 @@
 package org.jivesoftware.smackx.packet;
 
 import org.jivesoftware.smack.packet.IQ;
+import org.jivesoftware.smack.util.StringUtils;
 
 import java.util.Collections;
 import java.util.Iterator;
@@ -94,7 +95,7 @@
         buf.append("<query xmlns=\"http://jabber.org/protocol/disco#items\"");
         if (getNode() != null) {
             buf.append(" node=\"");
-            buf.append(getNode());
+            buf.append(StringUtils.escapeForXML(getNode()));
             buf.append("\"");
         }
         buf.append(">");
@@ -222,16 +223,16 @@
             StringBuilder buf = new StringBuilder();
             buf.append("<item jid=\"").append(entityID).append("\"");
             if (name != null) {
-                buf.append(" name=\"").append(name).append("\"");
+                buf.append(" name=\"").append(StringUtils.escapeForXML(name)).append("\"");
             }
             if (node != null) {
-                buf.append(" node=\"").append(node).append("\"");
+                buf.append(" node=\"").append(StringUtils.escapeForXML(node)).append("\"");
             }
             if (action != null) {
-                buf.append(" action=\"").append(action).append("\"");
+                buf.append(" action=\"").append(StringUtils.escapeForXML(action)).append("\"");
             }
             buf.append("/>");
             return buf.toString();
         }
     }
-}
\ No newline at end of file
+}
