Index: source/org/jivesoftware/smackx/pubsub/Node.java
===================================================================
--- source/org/jivesoftware/smackx/pubsub/Node.java	(revision 12558)
+++ source/org/jivesoftware/smackx/pubsub/Node.java	(working copy)
@@ -515,8 +515,11 @@
 				{
 					List<PacketExtension> secondLevelList = ((EmbeddedPacketExtension)embedEvent).getExtensions();
 					
-					if (secondLevelList.size() > 0 && secondLevelList.get(0).getElementName().equals(secondElement))
+					// XEP-0060 allows no elements on second level  
+					if(secondLevelList.size() == 0)
 						return true;
+					else if (secondLevelList.size() > 0 && secondLevelList.get(0).getElementName().equals(secondElement))
+						return true;
 				}
 			}
 			return false;
